client/consensus/common/src/lib.rs

```diff
@@ -74,47 +83,100 @@ impl<B: BlockT> ParachainConsensus<B> for Box<dyn ParachainConsensus<B> + Send +
 /// This is used to set `block_import_params.fork_choice` to `false` as long as the block origin is
 /// not `NetworkInitialSync`. The best block for parachains is determined by the relay chain. Meaning
 /// we will update the best block, as it is included by the relay-chain.
-pub struct ParachainBlockImport<I>(I);
+pub struct ParachainBlockImport<Block: BlockT, BI, BE> {
+	inner: BI,
+	monitor: Option<SharedData<LevelMonitor<Block, BE>>>,
```

Comment:
Using a `Mutex` here should be much simpler. We don't need this data being locked from somewhere else.

Reply:
(Correct me if I overlooked something or I didn't get your suggestion)

What if the block import method is called concurrently by two different components (e.g. by the authoring and import queue)?

If I release the lock between the read (before inner import) and the write of the cache (after inner import), a race condition may happen.

Assume this scenario
- we have a monitor limit set to something less than the hard limit in the backend (lets say this is infinite for the sake of this example).
- two components sharing the block import object.
- both of them import a block
- both of them read that the num of elements per level is `limit-1`.
- they don't prune nothing from
- in the end they will both add the imported block to the cache
- resulting in a num of blocks per level equal to `limit+1`.

Obviously (in this scenario) **if the monitor limit is equal to the hard limit of the backend** (32) then the second `inner.block_import` will fail with the `TooManySiblingBlocks` error (thus the second element is not imported in the cache) but we don't want this error to be triggered.

Reply:
Yeah, good points :+1: I didn't thought of importing a build block and importing a block from the network.